osquery/events/darwin/fsevents.cpp

```diff
@@ -192,17 +194,23 @@ void FSEventsEventPublisher::Callback(
 }
 
 bool FSEventsEventPublisher::shouldFire(
-    const FSEventsSubscriptionContextRef& mc,
+    const FSEventsSubscriptionContextRef& sc,
     const FSEventsEventContextRef& ec) const {
-  // This is stopping us from getting events on links.
-  // If we need this feature later, this line will have to be updated to
-  // understand links.
-  ssize_t found = ec->path.find(mc->path);
-  if (found != 0) {
+  if (sc->recursive) {
+    // This is stopping us from getting events on links.
+    // If we need this feature later, this line will have to be updated to
+    // understand links.
+    ssize_t found = ec->path.find(sc->path);
+    if (found != 0) {
+      return false;
+    }
+  } else if (fnmatch((sc->path + "*").c_str(),
+                     ec->path.c_str(),
+                     FNM_PATHNAME | FNM_CASEFOLD) != 0) {
```

Comment:
This should address #1142, using `FNM_PATHNAME` a trailing `/` character past any wildcards "*" (and at this point our SQL "%"s have become "*"s) will fail the function matching. Double wildcards will have been resolved beforehand and each of the `sc->path` checks will be the most specific folder path. There's one last optimization possible by detecting double wildcards in the subscription/configuration parsing and setting the recursive flag.